### PR TITLE
Reanimated keyboard observation

### PR DESCRIPTION
**Pull request by savelichalex**

### CODE REVIEW

**Comment by savelichalex** [comment]
We can merge it somewhere, but unfortunately can't use it until we re-make input in chat, as `adjustResize` will stop work with this

File: Example/android/app/src/main/java/tonlabs/uikit/MainApplication.java [hunk 1]
```diff
@@ -53,7 +67,7 @@ public ReactNativeHost getReactNativeHost() {
   public void onCreate() {
     super.onCreate();
     SoLoader.init(this, /* native exopackage */ false);
-    initializeFlipper(this, getReactNativeHost().getReactInstanceManager());
+//    initializeFlipper(this, getReactNativeHost().getReactInstanceManager());
```
Comment (sertony) [inline Example/android/app/src/main/java/tonlabs/uikit/MainApplication.java]: Any problems with Flipper?

File: Example/ios/Quiver UI.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -464,9 +464,9 @@
 			);
 			inputPaths = (
 				"${PODS_ROOT}/Target Support Files/Pods-Quiver UI-Quiver UITests/Pods-Quiver UI-Quiver UITests-frameworks.sh",
-				"${PODS_XCFRAMEWORKS_BUILD_DIR}/double-conversion/double-conversion.framework/double-conversion",
-				"${PODS_XCFRAMEWORKS_BUILD_DIR}/OpenSSL/OpenSSL.framework/OpenSSL",
-				"${PODS_XCFRAMEWORKS_BUILD_DIR}/hermes/hermes.framework/hermes",
+				"${PODS_XCFRAMEWORKS_BUILD_DIR}/Flipper-DoubleConversion/double-conversion.framework/double-conversion",
+				"${PODS_XCFRAMEWORKS_BUILD_DIR}/OpenSSL-Universal/OpenSSL.framework/OpenSSL",
+				"${PODS_XCFRAMEWORKS_BUILD_DIR}/hermes-engine/hermes.framework/hermes",
```
Comment (sertony) [inline Example/ios/Quiver UI.xcodeproj/project.pbxproj:469]: Something looks wrong here... (as we used an older RN version)

File: packages/keyboard/android/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    fbjni
+    PATHS ${LIBRN_DIR}
+    NO_CMAKE_FIND_ROOT_PATH
+)
+# find_library(
+#     JSI_LIB
+#     jsi
+#     PATHS ${LIBRN_DIR}
+#     NO_CMAKE_FIND_ROOT_PATH
+# )
```
Comment (sertony) [inline packages/keyboard/android/CMakeLists.txt]: Maybe remove it if we don't use it?

=======
Also I'd like to have a short tutorial about the used links here.
For instance why do we have to deal with `REACT_NATIVE_JNI_LIB` and `FBJNI_LIB`? (while `lib-react-native-jsi` of `tonclient` doesn't do it)

File: packages/keyboard/android/build.gradle [hunk 4]
```diff
@@ -17,15 +50,182 @@ android {
...
+
+    def inputFile = new File(NODE_MODULES_DIR, '/react-native/package.json')
+    def json = new JsonSlurper().parseText(inputFile.text)
+    def reactNativeVersion = json.version as String
+    def (major, minor, patch) = reactNativeVersion.tokenize('.')
+
+    def jsEngine = FOR_HERMES ? "hermes" : "jsc"
+    def reaAAR = "${NODE_MODULES_DIR}/react-native-reanimated/android/react-native-reanimated-${minor}-${jsEngine}.aar"
+
+    extractJNI(files(rnAAR, reaAAR))
```
Comment (sertony) [inline packages/keyboard/android/build.gradle]: Need to tell me more how `extractJNI` & `extractAARHeader` work here.

And what if we start building Surf from Android source again (where no `.aar` from react-native is really used).

File: packages/keyboard/android/build.gradle [hunk 5]
```diff
@@ -1,11 +1,44 @@
...
 apply plugin: 'com.android.library'
 
 project.ext {
-    buildToolsVersion = rootProject.ext.has("buildToolsVersion") ? rootProject.ext.buildToolsVersion : '27.0.3'
-    minSdkVersion = rootProject.ext.has("minSdkVersion") ? rootProject.ext.minSdkVersion : 19
-    compileSdkVersion = rootProject.ext.has("compileSdkVersion") ? rootProject.ext.compileSdkVersion : 27
-    targetSdkVersion = rootProject.ext.has("targetSdkVersion") ? rootProject.ext.targetSdkVersion : 25
-    supportLibVersion = rootProject.ext.has("supportLibVersion") ? rootProject.ext.supportLibVersion  : '27.1.1'
+    buildToolsVersion = rootProject.ext.has("buildToolsVersion") ? rootProject.ext.buildToolsVersion : '30.0.2'
+    minSdkVersion = rootProject.ext.has("minSdkVersion") ? rootProject.ext.minSdkVersion : 21
```
Comment (sertony) [inline packages/keyboard/android/build.gradle]: You're right. We forgot about this place

```suggestion
    minSdkVersion = rootProject.ext.has("minSdkVersion") ? rootProject.ext.minSdkVersion : 24
```

File: packages/keyboard/android/build.gradle [hunk 6]
```diff
@@ -1,11 +1,44 @@
...
+    compileSdkVersion = rootProject.ext.has("compileSdkVersion") ? rootProject.ext.compileSdkVersion : 30
+    targetSdkVersion = rootProject.ext.has("targetSdkVersion") ? rootProject.ext.targetSdkVersion : 30
+    supportLibVersion = rootProject.ext.has("supportLibVersion") ? rootProject.ext.supportLibVersion : '27.1.1'
```
Comment (sertony) [inline packages/keyboard/android/build.gradle]:
```suggestion
    supportLibVersion = rootProject.ext.has("supportLibVersion") ? rootProject.ext.supportLibVersion : '30.0.2'
```

File: Example/android/app/src/main/java/tonlabs/uikit/MainApplication.java (hunk 1 above)
Comment (savelichalex) [reply Example/android/app/src/main/java/tonlabs/uikit/MainApplication.java]: Probably already resolved

File: Example/ios/Quiver UI.xcodeproj/project.pbxproj (hunk 2 above)
Comment (savelichalex) [reply Example/ios/Quiver UI.xcodeproj/project.pbxproj:469]: Xcode did it for me for quite a long time, this time I decided to commit it

File: packages/keyboard/android/CMakeLists.txt (hunk 3 above)
Comment (savelichalex) [reply packages/keyboard/android/CMakeLists.txt]: About `jni` - `tonclient` use it, they just do it with `prefab`, while we not (took it from `react-native-vision-camera`)

File: packages/keyboard/android/build.gradle (hunk 4 above)
Comment (savelichalex) [reply packages/keyboard/android/build.gradle]: 1. simple - it unzips it and places under `.cxx` folder (it contains `.so`), then `cmake` able to find library there

2. my understanding is that it doesn't matter, as every library produces `.aar` anyway (with `.so` in it) during gradle build.

File: Example/ios/Quiver UI.xcodeproj/project.pbxproj (hunk 2 above)
Comment (sertony) [reply Example/ios/Quiver UI.xcodeproj/project.pbxproj:469]: Hm... Interesting. It was like this before in my xcode, but once we've updated the RN version it has changed =\

Comment (sertony) [reply Example/ios/Quiver UI.xcodeproj/project.pbxproj:469]: Well, it happened to be connected to the legacy Flipper version installed on your machine.
At the moment my links are correct:
https://github.com/facebook/flipper/blob/ec9e3aeb6e60973ce668ff516fc3ba8d969ccd02/react-native/ReactNativeFlipperExample/ios/ReactNativeFlipperExample.xcodeproj/project.pbxproj#L400..L406

File: packages/keyboard/cpp/UIKitKeyboardModule.cpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    std::shared_ptr<MutableValue> bottomMutable = bottom.asObject(runtime).getHostObject<MutableValue>(runtime);
+
+    _keyboardFrameListener.get()->addFrameListener(uid.asNumber(), [=, &bottom](double keyboardTopPosition) {
+        // At the point we will be on UI thread
+        // therefore we must take runtime on the same thread,
+        // and it's the runtime that reanimated uses
+        jsi::Runtime &reanimatedUIRuntime = *_nativeReanimatedModule->runtime.get();
+        bottomMutable.get()->set(reanimatedUIRuntime,
+                                 jsi::PropNameID::forAscii(reanimatedUIRuntime, "value"),
+                                 jsi::Value(keyboardTopPosition));
```
Comment (sertony) [inline packages/keyboard/cpp/UIKitKeyboardModule.cpp]: That is awesome how you figured out the way `ShareableValue` works!

File: packages/keyboard/ios/UIKitKeyboardIosFrameListener.m [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+RCT_EXPORT_MODULE()
+
++ (BOOL)requiresMainQueueSetup {
+    return YES;
+}
+
+/**
+ * You might have a question: "Does Apple allow this?"
+ * And it seems so.
+ * I got it from https://github.com/JunyuKuang/KeyboardUtility/blob/master/KeyboardUtility/KeyboardFrameObserversManager.swift#L102-L127
```
Comment (sertony) [inline packages/keyboard/ios/UIKitKeyboardIosFrameListener.m]: I'm glad that haven't been changed for years :)

File: packages/keyboard/src/useKeyboardBottom.tsx [hunk 9]
```diff
@@ -0,0 +1,37 @@
+import * as React from 'react';
+import type Animated from 'react-native-reanimated';
+
+let uidCounter = 0;
+
+function makeMutableBottom(uid: number): Animated.SharedValue<number> {
+    // @ts-ignore
```
Comment (sertony) [inline packages/keyboard/src/useKeyboardBottom.tsx]: You might want to `declare global`

File: packages/keyboard/src/useKeyboardBottom.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+function clearMutableBottom(uid: number) {
+    // @ts-ignore
+    // eslint-disable-next-line no-underscore-dangle
+    global.__uikitKeyboard.removeFrameListener(uid);
+}
+
+export function useKeyboardBottom() {
+    // TODO: should it be 0 or insets.bottom?
```
Comment (sertony) [inline packages/keyboard/src/useKeyboardBottom.tsx]: Didn't get about the `insets.bottom`...

**Review by sertony — APPROVED** [review]
Approved, but I'd like to discuss it with @savelichalex once he is back

**Review by sertony — APPROVED** [review]

**Review by SAnatoliiS — APPROVED** [review]

**Review by sertony — APPROVED** [review]

File: casts/keyboard/src/useKeyboardBottomInset.tsx [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+function makeMutableBottom(uid: number) {
+    return getUIKitKeyboardModule().addFrameListener(uid);
+}
+
+function clearMutableBottom(uid: number) {
+    getUIKitKeyboardModule().removeFrameListener(uid);
+}
+
+export function useKeyboardBottomInset() {
+    // TODO: should it be 0 or insets.bottom?
```
Comment (sertony) [inline casts/keyboard/src/useKeyboardBottomInset.tsx]: I also wonder.. :)

**Review by sertony — APPROVED** [review]

**Pull request merged**